docs/fundamentals/networking/http/httpclient-migrate-from-httpwebrequest.md

```diff
@@ -0,0 +1,624 @@
+
+## Usage of HttpClient and HttpRequestMessage Properties
+
+When working with HttpClient in .NET, you have access to a variety of properties that allow you to configure and customize HTTP requests and responses. Understanding these properties can help you make the most of HttpClient and ensure that your application communicates efficiently and securely with web services.
+
+### Example: Usage of HttpRequestMessage properties
+
+Here's an example of how to use HttpClient and HttpRequestMessage together:
+
+```csharp
+var client = new HttpClient();
+
+var request = new HttpRequestMessage(HttpMethod.Post, "https://example.com"); // Method and RequestUri usage
+var request = new HttpRequestMessage() // Alternative way to set RequestUri and Method
+{
+    RequestUri = new Uri("https://example.com"),
+    Method = HttpMethod.Post
+};
+request.Headers.Add("Custom-Header", "value");
+request.Content = new StringContent("somestring");
+
+using var response = await client.SendAsync(request);
+var protocolVersion = response.RequestMessage.Version; // Fetch `ProtocolVersion`.
+```
+
+### Example: Fetch Redirected URI
+
+Here's an example of how to fetch redirected URI (Same as `HttpWebRequest.Address`):
+
+```csharp
+var client = new HttpClient();
+using var response = await client.GetAsync(uri);
+var redirectedUri = response.RequestMessage.RequestUri;
+```
+
+## Usage of SocketsHttpHandler and ConnectCallback
+
+The `ConnectCallback` property in `SocketsHttpHandler` allows developers to customize the process of establishing a TCP connection. This can be useful for scenarios where you need to control DNS resolution or apply specific socket options on the connection. By using `ConnectCallback`, you can intercept and modify the connection process before it is used by `HttpClient`.
+
+### Example: Binding IP Address to Socket
+
+In the old approach using `HttpWebRequest`, you might have used custom logic to bind a specific IP address to a socket. Here's how you can achieve similar functionality using `HttpClient` and `ConnectCallback`:
+
+**Old Code Using `HttpWebRequest`**:
+
+```csharp
+HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
+request.ServicePoint.BindIPEndPointDelegate = (servicePoint, remoteEndPoint, retryCount) =>
+{
+    // Bind to a specific IP address
+    IPAddress localAddress = IPAddress.Parse("192.168.1.100");
+    return new IPEndPoint(localAddress, 0);
+};
+using HttpWebResponse response = (HttpWebResponse)request.GetResponse();
+```
+
+**New Code Using `HttpClient` and `ConnectCallback`**:
+
+```csharp
+var handler = new SocketsHttpHandler
+{
+    ConnectCallback = async (context, cancellationToken) =>
+    {
+        // Bind to a specific IP address
+        IPAddress localAddress = IPAddress.Parse("192.168.1.100");
+        var socket = new Socket(localAddress.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
```

Comment:
You need a try-catch block to eventually dispose the socket if connect fails, see: https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.connectcallback?view=net-8.0#examples
Applies to other examples bellow as well.